src/insertItem.ts

```diff
@@ -0,0 +1,449 @@
+        let functionName = await this.ui.showInputBox({ prompt: "Name of function?" });
+        let functionDef = await this.getFunction();
+        await this.insertInObjectInternal(members, textEditor, functionDef, functionName, 4);
+    }
+
+    private async insertFunction(template: DeploymentTemplate, textEditor: vscode.TextEditor): Promise<boolean> {
+        let functions = this.getTemplateArrayPart(template, templateKeys.functions);
+        if (!functions) {
+            // tslint:disable-next-line:no-unsafe-any
+            return await this.insertFunctionAsTopLevel(template.topLevelValue, textEditor);
+        }
+        if (functions.length === 0) {
+            return await this.insertFunctionAsNamespace(functions, textEditor);
+        }
+        let namespace = Json.asObjectValue(functions.elements[0]);
+        if (!namespace) {
+            vscode.window.showErrorMessage('The first namespace in functions is not an object!');
+            return false;
+        }
+        let members = namespace.getPropertyValue("members");
+        if (!members) {
+            await this.insertFunctionAsMembers(namespace, textEditor);
+            return true;
+        }
+        let membersObject = Json.asObjectValue(members);
+        if (!membersObject) {
+            vscode.window.showErrorMessage('The first namespace in functions does not have members as an object!');
+            return false;
+        }
+        await this.insertFunctionAsFunction(membersObject, textEditor);
+        return true;
+    }
+
+    private async insertResource(template: DeploymentTemplate, textEditor: vscode.TextEditor): Promise<boolean> {
+        let resources = this.getTemplateArrayPart(template, templateKeys.resources);
+        let pos: vscode.Position;
+        let index: number;
+        let text = "\r\n\t\t\r\n\t";
+        if (!resources) {
+            if (!template.topLevelValue) {
+                vscode.window.showErrorMessage("Invalid ARM template!");
+                return false;
+            }
+            // tslint:disable-next-line:no-any
+            let subPart: any = [];
+            index = await this.insertInObjectInternal(template.topLevelValue, textEditor, subPart, "resources", 1);
+            pos = textEditor.selection.active;
+        } else {
+            index = resources.span.endIndex;
+            if (resources.elements.length > 0) {
+                let lastIndex = resources.elements.length - 1;
+                index = resources.elements[lastIndex].span.afterEndIndex;
+                text = `,\r\n\t\t`;
+            }
+        }
+        const resource = await this.ui.showQuickPick(getResourceSnippets(), { placeHolder: 'What resource do you want to insert?' });
+        await this.insertText(textEditor, index, text, false);
+        let range = new vscode.Range(textEditor.document.positionAt(index), textEditor.document.positionAt(index + this.formatText(text, textEditor).length));
```

Comment:
This code needs some commenting about what it's trying to do

Reply:
I extracted a function and renamed some variables. 
Do you want comments too?